contracts/AutonomousDripper.sol

```diff
@@ -0,0 +1,187 @@
+    address private _keeperRegistryAddress;
+
+    constructor(
+        address beneficiaryAddress,
+        uint64 startTimestamp,
+        uint64 durationSeconds,
+        uint intervalSeconds,
+        address[] memory watchlistAddresses,
+        address keeperRegistryAddress
+    ) VestingWallet(
+        beneficiaryAddress,
+        startTimestamp,
+        durationSeconds
+    ) ConfirmedOwner(
+        msg.sender
+    ) {
+        lastTimestamp = startTimestamp;
+        interval = intervalSeconds;
+        assetsWatchlist = watchlistAddresses;
+        setKeeperRegistryAddress(keeperRegistryAddress);
+    }
+
+    /**
+     * @dev Setter for the list of ERC-20 token addresses to consider for
+     * releasing. Can only be called by the current owner.
+     */
+    function setAssetsWatchlist(address[] calldata newAssetsWatchlist) public virtual onlyOwner {
+        assetsWatchlist = newAssetsWatchlist;
+    }
+
+    /**
+     * @dev Loop over the assetsWatchlist and check their local balance.
+     * Returns a filtered version of the assetsWatchlist for which the local
+     * balance is greater than zero.
+     */
+    function _getAssetsHeld() internal view returns (address[] memory) {
+        address[] memory _assetsHeld = new address[](assetsWatchlist.length);
+        uint256 count = 0;
+        for (uint idx = 0; idx < assetsWatchlist.length; idx++) {
+            uint256 balance = IERC20(assetsWatchlist[idx]).balanceOf(address(this));
+            if (balance > 0) {
+                _assetsHeld[count] = assetsWatchlist[idx];
+                count++;
+            }
+        }
+        if (count != assetsWatchlist.length) {
+            assembly {
+                mstore(_assetsHeld, count)
+            }
+        }
+        return _assetsHeld;
+    }
+
+    /**
+     * @dev Runs off-chain at every block to determine if the `performUpkeep`
+     * function should be called on-chain.
+     */
+    function checkUpkeep(bytes calldata) external view override whenNotPaused returns (
+        bool upkeepNeeded, bytes memory checkData
+    ) {
+        if ((block.timestamp - lastTimestamp) > interval) {
+            address[] memory assetsHeld = _getAssetsHeld();
+            if (assetsHeld.length > 0) {
+                return (true, abi.encode(assetsHeld));
+            }
+        }
+    }
+
+    /**
+     * @dev Contains the logic that should be executed on-chain when
+     * `checkUpkeep` returns true.
+     */
+    function performUpkeep(bytes calldata performData) external override onlyKeeperRegistry whenNotPaused {
+        if ((block.timestamp - lastTimestamp) > interval) {
+            address[] memory assetsHeld = abi.decode(performData, (address[]));
+            for (uint idx = 0; idx < assetsHeld.length; idx++) {
+                if (IERC20(assetsHeld[idx]).balanceOf(address(this)) > 0) {
+                    VestingWallet.release(assetsHeld[idx]);
+                    lastTimestamp = block.timestamp;
```

Comment:
Can save gas by putting this after the loop (something like if(nonZero) -> lastTimestamp = block.timestamp

Reply:
thanks, implemented in 276fe10